File: Java/DFS/PathWithMaximumSum.java
/*
*   @author
*   Aakash Verma
*
*   Output:
*   42
*
*/

import java.util.*;

class Node {
    int data;
    Node left, right;
    public Node(int key) {
        data = key;
        left = right = null;
    }
}


class PathWithMaximumSum {

    static Node root;
    PathWithMaximumSum() {
        root = null;
    }

    private static int maxSum;

    private static int findMaxSum(Node root) {
        if(root == null) return 0;

        int leftSum = findMaxSum(root.left);
        int rightSum = findMaxSum(root.right);

        leftSum = Math.max(leftSum, 0);
        rightSum = Math.max(rightSum, 0);
        
        int currentSum = leftSum + rightSum + root.data;

        maxSum = Math.max(currentSum, maxSum);
        return Math.max(leftSum, rightSum) + root.data;
    }

    public static int findMaximumPathSum(Node root) {
        maxSum = Integer.MIN_VALUE;
        findMaxSum(root);
        return maxSum;
    }

    public static void main (String[] args) {
        root = new Node(10);
        root.left = new Node(2);
        root.right = new Node(10);
        root.left.left = new Node(20);
        root.left.right = new Node(1);
        root.right.left = new Node(-25);
        root.right.left.left = new Node(3);
        root.right.left.right = new Node(4);
        int ans = findMaximumPathSum(root);
        System.out.println(ans);
    }
}

File: Python/DFS/PathWithMaximumSum.py
# @author
# Aakash Verma

# Output:
# 42

from collections import *
import math

class Node:
    def __init__(self, data):
        self.data = data
        self.right = self.left = None

class PathWithMaximumSum:

    def __init__(self): 
        self.root = None

    def findMaxSum(self, root):
        if root is None: return 0

        leftSum = self.findMaxSum(root.left)
        rightSum = self.findMaxSum(root.right)

        leftSum = max(leftSum, 0)
        rightSum = max(rightSum, 0)

        currentSum = leftSum + rightSum + root.data

        self.maxSum = max(currentSum, self.maxSum)

        return max(leftSum, rightSum) + root.data

    def findMaximumPathSum(self, root):
        self.maxSum = -math.inf
        self.findMaxSum(root)
        return self.maxSum
        
        
if __name__ == '__main__':
    tree = PathWithMaximumSum()
    tree.root = Node(10)
    tree.root.left = Node(2)
    tree.root.right = Node(10)
    tree.root.left.left = Node(20)
    tree.root.left.right = Node(1)
    tree.root.right.left = Node(-25)
    tree.root.right.left.left = Node(3)
    tree.root.right.left.right = Node(4)
    ans = tree.findMaximumPathSum(tree.root)
    print(ans)


